File: gpt-4/template_context_full/acc_hostptr.c
#include <stdlib.h>
#include <openacc.h>
#include "acc_testsuite.h"

#ifndef T1
/*T1:acc hostptr, V:2.0-2.7*/

int test1() {

    int err = 0;
    srand(SEED);
    float *a;
    float *a_d;
    int i = 0;

    a = (float *) malloc(4 * sizeof(float));
    for(i=0; i<4; i++) {
        a[i] = (float)i;
    }

    /* Copy the host memory to the device */
    a_d = acc_copyin(a, 4 * sizeof(float));

    /* Test acc_hostptr */
    if(a != acc_hostptr(a_d)) {
        err = 1;
    }

    /* Cleanup */
    acc_delete(a, 4 * sizeof(float));
    free(a);
    
    return err;
}
#endif

int main() {
    int failcode = 0;
    int failed;

#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x) {
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif

    return failcode;
}

File: bad-tests-v2/acc_hostptr.f90
#ifndef T1
!T1:acc hostptr,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        TYPE(C_PTR) :: host_ptr, device_ptr
        REAL, POINTER :: host_array(:)
        INTEGER :: i

        ALLOCATE(host_array(10))
        DO i = 1, 10
          host_array(i) = i
        END DO

        device_ptr = acc_copyin(host_array, SIZEOF(host_array))
        host_ptr = acc_hostptr(device_ptr)

        IF (C_ASSOCIATED(host_ptr)) THEN
          IF (C_LOC(host_array) .NE. C_LOC(host_ptr)) THEN
            errors = errors + 1
          END IF
        ELSE
          errors = errors + 1
        END IF

        CALL acc_delete(device_ptr, SIZEOF(host_array))
        DEALLOCATE(host_array)

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END
#endif

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
#ifndef T1
        LOGICAL :: test1
#endif
        failed = .FALSE.
        failcode = 0
#ifndef T1
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
#endif
        CALL EXIT (failcode)
      END PROGRAM